tests/test_api.py

```diff
@@ -786,3 +786,163 @@ def test_product_components(client, api_path):
+        parent=dep_provide_node,
+        purl=dep2_comp.purl,
+        defaults={"obj": dep2_comp},
+    )
+
+    dep2_comp.save_component_taxonomy()
+    dep_comp.save_component_taxonomy()
+    root_comp.save_component_taxonomy()
+    root_comp.save()
+
+    assert dep_comp.purl in root_comp.provides.values_list("purl", flat=True)
+    assert dep2_comp.purl in root_comp.provides.values_list("purl", flat=True)
+
+    response = client.get(f"{api_path}/components?namespace=REDHAT")
+    assert response.status_code == 200
+    assert response.json()["count"] == 3
+
+    response = client.get(f"{api_path}/components/{root_comp.uuid}")
+    assert response.status_code == 200
+    assert response.json()["name"] == root_comp.name
+    assert response.json()["related_url"] == root_comp.related_url
+    assert len(response.json()["sources"]) == 0
+    assert len(response.json()["provides"]) == 2
+    assert len(response.json()["upstreams"]) == 1
+
+    response = client.get(f"{api_path}/components/{dep_comp.uuid}")
+    assert response.status_code == 200
+    assert len(response.json()["sources"]) == 1
+    assert len(response.json()["provides"]) == 1
+    assert len(response.json()["upstreams"]) == 1
+
+    response = client.get(f"{api_path}/components/{dep2_comp.uuid}")
+    assert response.status_code == 200
+    assert len(response.json()["sources"]) == 2
+    assert len(response.json()["provides"]) == 0
+    assert len(response.json()["upstreams"]) == 1
+
+    # retrieve all sources of root_comp
+    response = client.get(f"{api_path}/components?provides={root_comp.purl}")
+    assert response.status_code == 200
+    assert response.json()["results"] == []
+    assert response.json()["count"] == 0
+    # retrieve all provides of root_comp
+    response = client.get(f"{api_path}/components?sources={root_comp.purl}")
+    assert response.status_code == 200
+    assert response.json()["count"] == 2
+
+    # retrieve all sources of dep_comp component
+    response = client.get(f"{api_path}/components?provides={dep_comp.purl}")
+    assert response.status_code == 200
+    assert response.json()["count"] == 1
+    # retrieve all provides of dep_comp
+    response = client.get(f"{api_path}/components?sources={dep_comp.purl}")
+    assert response.status_code == 200
+    assert response.json()["count"] == 1
+
+    # retrieve all sources of dep2_comp component
+    response = client.get(f"{api_path}/components?provides={dep2_comp.purl}")
+    assert response.status_code == 200
+    assert response.json()["count"] == 2
+    # retrieve all provides of dep2_comp
+    response = client.get(f"{api_path}/components?sources={dep2_comp.purl}")
+    assert response.status_code == 200
+    assert response.json()["count"] == 0
+
+    # retrieve all components with upstream_comp upstream
+    response = client.get(f"{api_path}/components?upstreams={upstream_comp.purl}")
+    assert response.status_code == 200
+    assert response.json()["count"] == 3
```

Comment:
I think this was correct in the first version I reviewed, unless I've misunderstood. Today, only the root component reports an upstream component.
```suggestion
    assert response.json()["count"] == 1
```

Reply:
yes just investigating why  ...